portal-impl/src/com/liferay/portal/util/PortalImpl.java

```diff
@@ -5290,6 +5290,60 @@ public String getUniqueElementId(
 	}
 
 	@Override
+	public HttpServletRequest getUnwrappedDynamicServletRequest(
+		HttpServletRequest request) {
+
+		List<PersistentHttpServletRequestWrapper>
+			persistentHttpServletRequestWrappers = new ArrayList<>();
+
+		HttpServletRequestWrapper httpServletRequestWrapper = null;
+
+		HttpServletRequest currentRequest = request;
+
+		while (currentRequest instanceof HttpServletRequestWrapper) {
+			httpServletRequestWrapper =
+				(HttpServletRequestWrapper)currentRequest;
+
+			if (currentRequest instanceof DynamicServletRequest) {
+				return (HttpServletRequest)httpServletRequestWrapper.
+					getRequest();
+			}
+
+			if (currentRequest instanceof
+					PersistentHttpServletRequestWrapper) {
+
+				PersistentHttpServletRequestWrapper
+					persistentHttpServletRequestWrapper =
+						(PersistentHttpServletRequestWrapper)currentRequest;
+
+				persistentHttpServletRequestWrappers.add(
+					persistentHttpServletRequestWrapper.clone());
+			}
+
+			currentRequest =
+				(HttpServletRequest)httpServletRequestWrapper.getRequest();
+		}
+
+		if (ServerDetector.isWebLogic()) {
+			currentRequest = new NonSerializableObjectRequestWrapper(
+				currentRequest);
+		}
+
+		for (int i = persistentHttpServletRequestWrappers.size() - 1; i >= 0;
+			 i--) {
+
+			httpServletRequestWrapper =
+				persistentHttpServletRequestWrappers.get(i);
+
+			httpServletRequestWrapper.setRequest(currentRequest);
+
+			currentRequest = httpServletRequestWrapper;
+		}
+
+		return (HttpServletRequest)httpServletRequestWrapper.getRequest();
```

Comment:
Shouldn't this be using "currentRequest" instead of "httpServletRequestWrapper"? Seems like this could produce an NPE or return the incorrect request.